compiler/luci-interpreter/src/kernels/Mul.test.cpp

```diff
@@ -78,6 +78,68 @@ TEST(MulTest, Float)
   }
 }
 
+TEST(MulTest, SInt16)
+{
+  Shape base_shape = {2, 3, 1, 2};
+  std::vector<Shape> test_shapes{{1, 1, 3, 2}, {1, 3, 1, 2}, {2, 1, 3, 1}, {2, 3, 1, 1}};
+  std::vector<std::vector<int32_t>> ref_output_shapes{
+      {2, 3, 3, 2}, {2, 3, 1, 2}, {2, 3, 3, 2}, {2, 3, 1, 2}};
+
+  std::vector<float> input1_data{-0.3f, 2.3f, 0.9f,  0.5f, 0.8f, -1.1f,
+                                 1.2f,  2.8f, -1.6f, 0.0f, 0.7f, -2.2f};
+  std::vector<float> input2_data{0.2f, 0.3f, -0.4f, 0.5f, 1.0f, 0.9f};
+  std::vector<std::vector<float>> ref_outputs = {
+      {0.00f, 0.69f, 0.12f, 1.15f, 0.00f, 2.07f, 0.18f, 0.15f, 0.00f, 0.25f, 0.90f, 0.45f,
+       0.16f, 0.00f, 0.00f, 0.00f, 0.80f, 0.00f, 0.24f, 0.84f, 0.00f, 1.40f, 1.20f, 2.52f,
+       0.00f, 0.00f, 0.64f, 0.00f, 0.00f, 0.00f, 0.14f, 0.00f, 0.00f, 0.00f, 0.70f, 0.00f},
+      {0.00f, 0.69f, 0.00f, 0.25f, 0.80f, 0.00f, 0.24f, 0.84f, 0.64f, 0.00f, 0.70f, 0.00f},
+      {0.00f, 0.46f, 0.00f, 0.69f, 0.12f, 0.00f, 0.18f, 0.10f, 0.27f, 0.15f, 0.00f, 0.00f,
+       0.16f, 0.00f, 0.24f, 0.00f, 0.00f, 0.44f, 0.60f, 1.40f, 1.20f, 2.80f, 1.08f, 2.52f,
+       0.00f, 0.00f, 0.00f, 0.00f, 0.00f, 0.00f, 0.35f, 0.00f, 0.70f, 0.00f, 0.63f, 0.00f},
+      {0.00f, 0.46f, 0.27f, 0.15f, 0.00f, 0.44f, 0.60f, 1.40f, 0.00f, 0.00f, 0.63f, 0.00f}};
+  for (size_t i = 0; i < test_shapes.size(); ++i)
+  {
+    Tensor input1_tensor = makeInputTensor<DataType::S16>(base_shape, 3.0 / 32767, 0, input1_data);
+    Tensor input2_tensor =
+        makeInputTensor<DataType::S16>(test_shapes[i], 1.0 / 32767, 0, input2_data);
+    Tensor output_tensor = makeOutputTensor(DataType::S16, 4.0 / 32767, 0);
+
+    MulParams params{};
+    params.activation = Activation::RELU;
+
+    Mul kernel(&input1_tensor, &input2_tensor, &output_tensor, params);
+    kernel.configure();
+    kernel.execute();
+
+    EXPECT_THAT(extractTensorShape(output_tensor),
+                ::testing::ElementsAreArray(ref_output_shapes[i]))
+        << "With shape number " << i;
+    EXPECT_THAT(dequantizeTensorData(output_tensor), FloatArrayNear(ref_outputs[i], 0.001))
```

Comment:
I found the below code in TFLite.

```cpp
// For quantized Mul, the error shouldn't exceed (2*step + step^2).
...
const float kQuantizedStepInt16 = 2.0 / 32767.0;
const float kQuantizedToleranceInt16 =
    2.0 * kQuantizedStepInt16 + kQuantizedStepInt16 * kQuantizedStepInt16
```

How about using `2.0 * kQuantizedStepInt16 + kQuantizedStepInt16 * kQuantizedStepInt16` instead of `0.001` ?

Reply:
I just don't understand the formula :) I can also see another one in `TEST(QuantizedMulOpTest, NoActivationInt16Scaled)`, which is:
```C++
  const float kQuantizedToleranceInt16Scaled =
      6.0 * kQuantizedStepInt16 + kQuantizedStepInt16 * kQuantizedStepInt16;
```

Also, TFLite has its own way of quantizing input tensors, so I'm not sure these formulae will fit anyway.

But I think I can adjust scales to improve precision.

Reply:
@jinevening PTAL, does it looks better now? If it's OK, I will apply the same concept to `Add` in sibling PR.

Reply:
> Also, TFLite has its own way of quantizing input tensors, so I'm not sure these formulae will fit anyway.

I agree.

One heuristic we've used in uint8 was to use one step (=scale) of quantized values as the tolerance. In this example, the scale of output is 0.01, and tolerance is 1e-5 (in `FloatArrayNear`). So I think it is ok to use the current values.

(optional) Smaller scale would be better for testing precision (previous scales look better to me). How about using the previous scales and setting the tolerance to `n X scale of output tensor` (maybe `n=1` as uint8 quant)? Showing the relationship between scale and tolerance would improve readability.

Reply:
> (optional) Smaller scale would be better for testing precision (previous scales look better to me). How about using the previous scales and setting the tolerance to n X scale of output tensor (maybe n=1 as uint8 quant)? Showing the relationship between scale and tolerance would improve readability.

OK, I'll try.

Reply:
Done, PTAL